### ISSUE TITLE
[FR] TypeScriptReact syntax highlighting (tsx)

### ISSUE THREAD

**Issue by atdrago** [refined-bitbucket/refined-bitbucket#129]
Currently, only `.ts` extensions get syntax highlighting, whereas `.tsx` files get none. Support for `.tsx` files is built into `language-typescript`, so it seems like this would be an easy one to add here https://github.com/refined-bitbucket/refined-bitbucket/blob/dev/src/syntax-highlight/language-ext.js#L65

Is adding support for `.tsx` as simple as adding `'.tsx': 'language-typescript',` after the line referenced above? And if so, would you accept a PR for this feature?

Thank you for this extension! We just started using it on our team and love it so far!

**Comment by reyronald** [comment]
Definitely! Feel free to open the PR and I'll merge it right away!

Glad to hear that your team is enjoying the extension ! 😊 😊

**Comment by reyronald** [comment]
Closed via #130

**Comment by reyronald** [comment]
New [version 3.5.0](https://github.com/refined-bitbucket/refined-bitbucket/releases/tag/v3.5.0) is published with this fix.

Let me know if it works properly!

**Comment by atdrago** [comment]
Yep, working as expected! The only "issue" is that it lights up keywords that appear within text nodes of a JSX element. Definitely not a big deal though.
![screen shot 2018-02-01 at 11 05 07 am](https://user-images.githubusercontent.com/144225/35697929-6ed4be54-0740-11e8-82b4-1243dabe994b.png)

**Comment by reyronald** [comment]
Hmm, perhaps it would have been a better idea to use React JSX instead of TypeScript.

Could you copy / paste some of your code [here](http://prismjs.com/test.html), choose between those two and check which one works better?

**Comment by reyronald** [comment]
Seems like `language-typescript` doesn't actually support JSX out of the box, look at this currently open PR https://github.com/PrismJS/prism/pull/1277 which is adding it to PrismJS.

I'll follow on it and see if the highlighting improves once that one is merged

**Comment by atdrago** [comment]
Hmm... I think you're right... my apologies. React JSX doesn't solve the above issue, but does give some nicer coloring. Top is React JSX, bottom is TypeScript
![screen shot 2018-02-01 at 12 39 37 pm](https://user-images.githubusercontent.com/144225/35702141-20a15e2e-074d-11e8-96b3-8af9c7415e79.png)

**Comment by reyronald** [comment]
Yeah, we definitely jumped the gun.

I'll wait a little to see if the above PR for the official TSX support is merged before releasing a patch, let's hope it's better than the alternatives we have now!

**Comment by atdrago** [comment]
Agreed! 

I definitely should've researched a bit more. I saw `language-typescript` and [went here](https://atom.io/packages/language-typescript), which is what I use in Atom. That lead me to [this file](https://github.com/Microsoft/TypeScript-TmLanguage/blob/master/TypeScriptReact.tmLanguage). So `language-typescript` is used for syntax highlighting in Atom, but it's clear to me now it has nothing to do with Prism. Again, my apologies.

Please feel free to @ me if you need more examples or help testing! In the mean time, should this get opened back up?

**Comment by reyronald** [comment]
Hehe, don't sweat it, and yeah let's open it again.

Let's wait a few days to see if https://github.com/PrismJS/prism/pull/1277 gets merged so we can use that, in the mean time just tell your team to bare with it 😅 

EDIT: TSX support for PrismJS was move to another PR https://github.com/PrismJS/prism/pull/1280

**Comment by reyronald** [comment]
Hey @atdrago , the PR was merged and React TSX is now available in http://prismjs.com/test.html.

Please check there to see if that syntax highlighting works better for your code base and let me know!

**Comment by atdrago** [comment]
Hey @reyronald, this looks great! It's just slightly nicer than React JSX, which is to be expected, but I'm glad there's a proper solution. It's funny how all of this happened at the same time.

**Some screenshots**

Only real difference here is that the type of `string` is highlighted:
<img width="679" alt="screen shot 2018-02-05 at 6 34 43 am" src="https://user-images.githubusercontent.com/144225/35810094-6cc11ba6-0a3f-11e8-8cef-bbed7c9cdde8.png">
<img width="677" alt="screen shot 2018-02-05 at 6 34 28 am" src="https://user-images.githubusercontent.com/144225/35810096-6ceb59fc-0a3f-11e8-8452-1db0422bef97.png">

No difference between React JSX and TSX for this screenshot (JSX is on top). `this` is still highlighted (and I'm also noticing `!` now too), but again that's not a show-stopper obviously:
<img width="908" alt="screen shot 2018-02-05 at 6 46 59 am" src="https://user-images.githubusercontent.com/144225/35810379-69998f02-0a40-11e8-8285-e2576facfb2a.png">
<img width="913" alt="screen shot 2018-02-05 at 6 44 18 am" src="https://user-images.githubusercontent.com/144225/35810274-10f0b902-0a40-11e8-8673-55f2ef6af90d.png">

**Comment by reyronald** [comment]
Hehe yeah I thought the same thing, what a coincidence that both issues were opened at the same time! Extremely convenient 😆 

Like you said, it is definitely better and I think TSX users will be satisfied, although I'm noticing that some improvements could be made, for instance the `ComponentAProps` type in the `ComponentA` declaration (first line) should be highlighted, just like `string`, right? If I have some time I'll use that code sample to see if I can submit a PR to PrismJS to fix that.

Just to be sure, could you do a local test with the dev version of the extension? I don't have access to any PR in Bitbucket with TSX code so I can't validate myself right now if the update is going to work properly. It's extremely easy and should only take you a few minutes

1. Just clone this repo
1. Checkout the branch `update-prism`
1. Run `npm install`
1. Run  `npm run watch`
1. Go to `chrome://extensions`
1. Disable your actual Refined Bitbucket extension
1. Enable the Developer Mode
1. Click `Load unpacked extension...`
1. Browse to the `extension` folder in the repo root.
1. That's it, be sure to open the Options page to see if syntax highlighting is enabled.
1. Refresh your cache when you open Bitbucket again!

If you can't, I'll try to do it myself later but I'll need more time to set it up.

**Comment by atdrago** [comment]
I'll definitely be able to do this by this weekend, if not sooner! I'll keep you updated!

**Comment by atdrago** [comment]
Hey @reyronald. Sorry for the delay!

I was able to follow your instructions and load the extension locally. Here's what I found (I did some name changing inline because I'm paranoid):

Looks like the `<ComponentProps>` is getting highlighted now!  I don't think that was happening before. However, types specified after a colon still don't get highlighted (like `Option[]` or the second `ComponentProps`)

<img width="583" alt="screen shot 2018-02-10 at 8 25 28 am" src="https://user-images.githubusercontent.com/144225/36064255-28629ee2-0e3d-11e8-939a-0913fb2f8492.png">

I took this just to show that my custom component `RadioPillBox` does not get highlighted:

<img width="500" alt="screen shot 2018-02-10 at 8 27 42 am" src="https://user-images.githubusercontent.com/144225/36064254-284a4158-0e3d-11e8-99db-f44c3291468b.png">

**Comment by reyronald** [comment]
Hey @atdrago , thank you so much for taking the time to test it! I know it can be difficult to find time to invest on this volunteer-type of work! 🌟 

To the extension's concern, what we want to know is if the new language definition that was introduced in the new version of Prism that I updated in that branch is loaded properly, instead of the old one we had. To verify this, we just need to know if the highlighting for those files in that branch is improved over the version that's released in the Webstore.

If you are not completely sure, you can easily switch between both extensions that you have installed (the development one and the released one) by just disabling one and re-enabling the other,  refreshing Bitbucket and try to spot the differences. If the one on this branch is different (and hopefully better) then our work is done. By the way, some times it is necessary to refresh the cache after switching, keep that in mind!

The improvements you mention are totally valid, but those belong to PrismJS itself, since that's the library I am using here to do the hard-work of highlighting things. If you feel like it, you could totally open up an issue in their repo with these suggestions and maybe they can take a look at it ! It would be great if that happens.

Again, thank you for your availability with this ! 🙌

**Comment by atdrago** [comment]
> If I have some time I'll use that code sample to see if I can submit a PR to PrismJS to fix that.

Apologies, man! I read this and thought you wanted me to tell you how the TS highlighting could be improved. 🤭

I went back and checked again. I loaded the extension from your branch in one window, then opened a different window, disabled the extension from your branch and re-enabled the one from the store, then cleared my cache and loaded up a PR in that window. Refreshed a few times just to be sure, but your version is definitely an improvement and preferable over 3.5.0. It includes the fix I mentioned above with `Component<ComponentProps>`, and also added highlighting for elements within the JSX. You can see in the screenshots above that `<React.Fragment>` and `<div>` get highlighted. In 3.5.0, elements aren't getting highlighted at all. (The issue I mentioned with `<RadioPillBox` above I think is actually because the attributes are not on the same line, but that's neither here nor there. 🙃)

So, to answer your question, it looks like the new version of PrismJS is loaded correctly! Thanks again so much for your work on this! 😄

**Comment by reyronald** [comment]
Perfect!! I'm glad to hear that!

Give me just a few days and I'll publish a new version with this fix! I'll inclulde you as a co-author on the pull request if you don't mind since you were involved in testing and verifying this! Which I appreciate a lot!

> (The issue I mentioned with <RadioPillBox above I think is actually because the attributes are not on the same line, but that's neither here nor there. 🙃)

Yeah, actually I have the same problem with JSX in my code base and I've noticed that's definitely the reason. I just learned to live with it 😅 , but maybe it can be fixed upstream in PrismJS, I'll ask them later when I have some time

**Comment by reyronald** [comment]
Closed via #137

**Comment by reyronald** [comment]
Out in v3.6.0! 

Let me know if everything is working as expected in the released version!

**Comment by atdrago** [comment]
Looks great, @reyronald! Thanks again! 😊
